torch/_dynamo/output_graph.py

```diff
@@ -2229,6 +2199,296 @@ def maybe_lift_tracked_freevar_to_input(self, arg):
             return arg
         return self.lift_tracked_freevar_to_input(arg)
 
+    # See NOTE: [Auto lift basic free symbols when create_graph_input] for overall design
+    # This function should be used to track the unbacked symbols with its proxy created during
+    # dynamo tracing. It's invoked in wrap_fx_proxy for every proxied call for tensors and symbols.
+    # So that subgraph knows how to bind a symbol input with parent's proxy.
+    def lazy_bind_unbacked_symbols(self, example_value, e_proxy: torch.fx.Proxy):
+        # When binding the symbols in an exmaple_value, we bind the symbols
+        # to the proxy's associatied Tracer instead of current tracer.
+        # This is because:
+        # 1. We maybe calling wrap_tensors during speculate_subgraph because
+        # the variables are lazily realized. The proxy are top-level phs but
+        # current tracer is a subtracer.
+        # 2. For autograd.Function, we trace the backward graph with a new tracer
+        # whose parent is the forward tracer, but we're using all the proxies created
+        # in forward tracer to trace the backward.
+        # For example, forward calls save_for_backward for a input tensor t.
+        # Backward calls t.tolist(). In this case, all the proxies that backward tracer
+        # sees are from parent tracer (i.e. the forward tracer). (e.g. t[0].item())
+        # See test_validate_outputs_unbacked for repro on 2.
+        assert isinstance(e_proxy.tracer, SubgraphTracer)
+        if isinstance(example_value, torch.Tensor):
+            self._lazy_bind_unbacked_symbols_in_tensors(
+                e_proxy.tracer, example_value, e_proxy
+            )
+        elif isinstance(example_value, torch.SymInt):
+            e_proxy.tracer.bound_symbols[example_value.node.expr] = e_proxy
+            set_example_value(e_proxy.node, example_value)
+
+    # Unbacked symbols may be created as a function call in parent (e.g. t.nonzero).
+    # We use a lazy proxy for tracking, so that only when the unbacked symbol is actually used,
+    # we create the proxy for it. The exact use case is when a child subgraph looks for the proxy
+    # of an unbacked Symbol created in parent graph, we create a proxy in parent graph so
+    # that the subgraph bind the lifted symbol to subgraph.
+    def _lazy_bind_unbacked_symbols_in_tensors(
```

Comment:
There's some code duplication with "_lift_symbols_in_tensor". It sounds like you want a shared API for "iterating over all symbols in a tensor", that you can pass a function to.

Reply:
I tried this refacotering but it seems make the the code more complicated. Something like P1674013782 (couldn't finish). The interface looks like: 


The tricky part is that the lambda fn also need to know whether we're dealing with a size call or a stride call (to create sym_size or sym_stride) and also needs to deal with logging, which all needs to parametrized to use the shared API. Also for lift_symbols, we require the src but for track_unbacked_symbols, we require proxy, which means we need different ways of creating the proxies, which makes the callbacks very complicated ( The approach I tried is to rely on source to construct the proxy) but sometimes the src can be None, which means we cannot construct source but still need to record the relationship. Also we need to pass two callbacks, that one handles_symint_callback and handle_tensor_callback. All these complexity seems outweighs the benefit we have of the code redundancy...